### PR TITLE
Profile `LightningDataModule` hooks

### LINKED ISSUE #12967: Profile `LightningDataModule` hooks
## 🚀 Feature

Profile `LightningDataModule` hooks just like that of `LightningModule`.

Is there any specific reason why they are not profiled? Or it's just missed?

### Motivation

some hooks like `setup`, `prepare_data` when defined under `LightningModule` are profiled but when defined under `LightningDataModule` are not. Ideally they should be profiled too.

https://github.com/PyTorchLightning/pytorch-lightning/blob/5641836b96a6e85f1939e4d845d470931608e568/pytorch_lightning/trainer/trainer.py#L1447-L1450

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @carmocca @kaushikb11 @ninginthecloud @rohitgr7 @awaelchli @ananthsub @jjenniferdai

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #12967

### Does your PR introduce any breaking changes? If yes, please list them.
None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @kaushikb11 @ninginthecloud @rohitgr7 @awaelchli @ananthsub @jjenniferdai

### CODE REVIEW

**Comment by rohitgr7** [comment]
not sure what should the best minimal test to validate the change. Any recommendations?

Thank you :)

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1555,6 +1555,27 @@ def _call_lightning_module_hook(
 
         return output
 
+    def _call_lightning_datamodule_hook(
+        self,
+        hook_name: str,
+        *args: Any,
+        datamodule: Optional["pl.LightningDataModule"] = None,
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:1562]: We don't need this argument. It's in `_call_lightning_module_hook` for legacy reasons

```suggestion
```

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1555,6 +1555,27 @@ def _call_lightning_module_hook(
...
+            raise TypeError("No `LightningDataModule` is available to call hooks on.")
+
+        fn = getattr(datamodule, hook_name)
+        if not callable(fn):
+            return
+
+        with self.profiler.profile(f"[LightningDataModule]{datamodule.__class__.__name__}.{hook_name}"):
+            output = fn(*args, **kwargs)
+
+        return output
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:1577]:
```suggestion
        if callable(fn):
            with self.profiler.profile(f"[LightningDataModule]{datamodule.__class__.__name__}.{hook_name}"):
                return fn(*args, **kwargs)
```

**Review by carmocca — APPROVED** [review]

**Comment by carmocca** [comment]
> not sure what should the best minimal test to validate the change. Any recommendations?

I added a test like that in https://github.com/PyTorchLightning/pytorch-lightning/pull/12124 if you want to copy the pattern

**Review by akihironitta — APPROVED** [review]
🚀

File: tests/core/test_datamodules.py [hunk 3]
```diff
@@ -491,3 +492,62 @@ class BoringDataModule2(LightningDataModule):
     assert hasattr(BoringDataModule2, "__repr__")
     assert BoringDataModule2(batch_size=32).prepare_data() is None
     assert BoringDataModule2(batch_size=32) == BoringDataModule2(batch_size=32)
+
+
+@RunIf(skip_windows=True)  # TODO: all durations are 0 on Windows
```
[the remark this answers is not in the document]
Comment (carmocca) [reply tests/core/test_datamodules.py:497]: I believe so: #12998

[the remark this answers is not in the document]
Comment (rohitgr7) [reply tests/core/test_datamodules.py:497]: https://github.com/PyTorchLightning/pytorch-lightning/issues/12998

we don't check summaries for all
